### ISSUE TITLE
Add discriminants on enum variants

### ISSUE THREAD

**Issue by feds01** [hash-org/hashc#993]
This feature will allow users to specify the values of discriminants in enums. The proposed syntax is the following:
```rs
Kind := enum(
      A = 1,
      B = 2,
      C = 3,
      D = 4
      ...
)
```

The constraints on discriminants are:
- you are not allowed to specify them on enum variants that have associated fields, e.g. you cannot write something like:
```rs
Kind := enum(
      A = 1,
      B(u32, f32) = 2,
      ...
)

```
- discriminant values must be unique.
- The value of the discriminant cannot exceed the discriminant type, i.e. if the type is specified as `u8`, a discriminant cannot be larger than `255`.

**Comment by kontheocharis** [comment]
Why not allow discriminants with the associated fields? Additionally what is the purpose of specifying explicitly the type of the discriminant? Would it not make sense to always use the minimum integer size that fits the number of variants?

**Comment by feds01** [comment]
It might be for FFI reasons. For example, LLVM can define an enum like
```rs
Attribute := enum(
   Cold,
   Recursive,
   ReadOnly,
   ...
)
```
However, it defines the representation of the enum with a `u32`. If we were to always use the minimum integer size, this would create an incompatibility. Whilst the default behaviour should be to use the smallest possible integer (with respect to alignment/padding). So, if the user specifies a representation, then it should be respected and applied for these kinds of reasons.

**Comment by kontheocharis** [comment]
Yeah makes sense that you can specify it manually, though surely by default you shouldn't have to specify anything right?

**Comment by feds01** [comment]
Another thought, I don't see any particular reason to actually disallow something like:
```rs
Kind := enum(
      A = 1,
      B(u32, f32) = 2,
      ...
)
```

**Comment by feds01** [comment]
> Yeah makes sense that you can specify it manually, though surely by default you shouldn't have to specify anything right?

Yeah, I think you shouldn't need to specify the type of the discriminant, it should try to attempt to use the smallest discriminant variant. However, there is a slight issue from the point of the TC, given that it will try to "bake" the literals, what is the default type that it assumes the discriminant is?

**Comment by feds01** [comment]
I guess we could assume the default discriminant type is something like `isize` or `usize`, and then we can attempt to reduce it to the smallest possible integer variant?

**Comment by feds01** [comment]
Another question, it might be useful for there to be the ability to cast between an enum to the type of the enum tag or to access the tag of an enum value. Perhaps it should be an intrinsic, or even two intrinsics:
```rs
discriminant_type := <T, U> => (T) -> U

discriminant_value := <T, U> => (T) -> U
```
or perhaps  there should be some way to directly ask what the discriminant is on the type itself, i.e.
```rs
Kind := enum(
    A = 1,
    B = 2
)

foo := (item: Kind) => {
   discriminant := item._discriminant // or .discriminant() or a prelude function discriminant_of()
}
```

**Comment by kontheocharis** [comment]
I mean, in terms of enum variants with data, yes you could also specify the discriminants there, though you won't be able to just transmute one into the other. In terms of the second comment, going off of the syntax in #955, there could be some trait automatically implemented by enums:

```rs
IsEnum := struct <E> (
    variant_count: usize,
    discriminant: E -> 0..<variant_count
)
```

**Comment by kontheocharis** [comment]
Well the default type should just be the minimum numeric type that fits the number of variants surely..

**Comment by feds01** [comment]
As I explained, it has to first bake it into something, so what should the default type be?

**Comment by kontheocharis** [comment]
I don't see why TC cannot just look up the tag type...

**Comment by feds01** [comment]
Im not sure what that means, the tag type is determined at TC through the following context:
- Either there is a `#[repr(uXXX|iXXX)]` annotation
- or it defaults to some default tag value

Hence, I am asking what should be the default tag value?

**Comment by kontheocharis** [comment]
It must look at all the given discriminants, then do `let max = discriminants.iter().max()` and then determine what to bake it as through `max`. So it short-circuits the default integer baking on the case that it appears in an enum.
